plugins/modules/ec2_vpc_dhcp_option.py

```diff
@@ -190,91 +245,233 @@
 
 """
 
-import collections
-from time import sleep, time
-
 try:
-    import boto.vpc
-    import boto.ec2
-    from boto.exception import EC2ResponseError
+    import botocore
 except ImportError:
-    pass  # Taken care of by ec2.HAS_BOTO
+    pass  # Handled by AnsibleAWSModule
 
 from ..module_utils.core import AnsibleAWSModule
-from ..module_utils.ec2 import HAS_BOTO
-from ..module_utils.ec2 import connect_to_aws
-from ..module_utils.ec2 import get_aws_connection_info
+from ..module_utils.core import is_boto3_error_code
+from ..module_utils.ec2 import AWSRetry
+from ..module_utils.ec2 import ansible_dict_to_boto3_tag_list
+from ..module_utils.ec2 import boto3_tag_list_to_ansible_dict
+from ..module_utils.ec2 import camel_dict_to_snake_dict
+from ..module_utils.ec2 import compare_aws_tags
+from ..module_utils.ec2 import normalize_ec2_vpc_dhcp_config
 
 
-def get_resource_tags(vpc_conn, resource_id):
-    return dict((t.name, t.value) for t in vpc_conn.get_all_tags(filters={'resource-id': resource_id}))
+def ensure_tags(client, module, dhcp_options_id, tags, purge_tags):
+    changed = False
+    tags_to_unset = False
+    tags_to_set = False
 
+    if module.check_mode and dhcp_options_id is None:
+        # We can't describe tags without an option id, we might get here when creating a new option set in check_mode
+        return changed
 
-def retry_not_found(to_call, *args, **kwargs):
-    start_time = time()
-    while time() < start_time + 300:
-        try:
-            return to_call(*args, **kwargs)
-        except EC2ResponseError as e:
-            if e.error_code in ['InvalidDhcpOptionID.NotFound', 'InvalidDhcpOptionsID.NotFound']:
-                sleep(3)
-                continue
-            raise e
+    current_tags = boto3_tag_list_to_ansible_dict(client.describe_tags(aws_retry=True, Filters=[{'Name': 'resource-id', 'Values': [dhcp_options_id]}])['Tags'])
+
+    if tags:
+        tags_to_set, tags_to_unset = compare_aws_tags(current_tags, tags, purge_tags=purge_tags)
+    if purge_tags and not tags:
+        tags_to_unset = current_tags
+
+    if tags_to_unset:
```

Comment:
as fas as I see `if tags_to_unset:` is not necessary here. The logic can instantly applied after L276 `if purge_tags and not tags`  
but on the other hand, the code logic looks cleaner and more separated.

Reply:
My concern is keeping things separated enough for the changed logic to be clear.  I'm not sure where I would like to put the changed=True to cover a case where there are no tags_to_set but there are tags_to_unset without potentially setting the wrong changed in other cases.  It might be a bit verbose this way but it's consistent with what we do in other modules, my inclination would be to leave it a bit more easy to read.